libiberty/d-demangle.c

```diff
@@ -1468,66 +1590,53 @@ dlang_parse_qualified (string *decl, const char *mangled,
       while (*mangled == '0')
 	mangled++;
 
-      mangled = dlang_identifier (decl, mangled, options, kind);
+      mangled = dlang_identifier (decl, mangled, info);
 
       /* Consume the encoded arguments.  However if this is not followed by the
-	 next encoded length, then this is not a continuation of a qualified
-	 name, in which case we backtrack and return the current unconsumed
-	 position of the mangled decl.  */
+	 next encoded length or mangle type, then this is not a continuation of
+	 a qualified name, in which case we backtrack and return the current
+	 unconsumed position of the mangled decl.  */
       if (mangled && (*mangled == 'M' || dlang_call_convention_p (mangled)))
 	{
+	  string mods;
 	  const char *start = mangled;
 	  int saved = string_length (decl);
-	  string args;
+
+	  /* Save the type modifiers for appending at the end if needed.  */
+	  string_init (&mods);
 
 	  /* Skip over 'this' parameter and type modifiers.  */
 	  if (*mangled == 'M')
 	    {
 	      mangled++;
-	      mangled = dlang_type_modifiers (decl, mangled);
+	      mangled = dlang_type_modifiers (&mods, mangled);
 	      string_setlength (decl, saved);
 	    }
 
-	  /* The rule we expect to match in the mangled string is:
-
-		TypeFunctionNoReturn:
-		    CallConvention FuncAttrs Arguments ArgClose
+	  mangled = dlang_function_type_noreturn (decl, NULL, NULL,
+						  mangled, info);
+	  if (suffix_modifiers)
+	    string_appendn (decl, mods.b, string_length (&mods));
 
-	     The calling convention and function attributes are not included
-	     in the demangled string.  */
-	  mangled = dlang_call_convention (decl, mangled);
-	  mangled = dlang_attributes (decl, mangled);
-	  string_setlength (decl, saved);
-
-	  string_init (&args);
-	  mangled = dlang_function_args (&args, mangled, options);
-
-	  if (options & DMGL_PARAMS)
-	    {
-	      string_append (decl, "(");
-	      string_appendn (decl, args.b, string_length (&args));
-	      string_append (decl, ")");
-	    }
-
-	  string_delete (&args);
-
-	  if (mangled == NULL || !ISDIGIT (*mangled))
+	  if (mangled == NULL || *mangled == '\0')
```

Comment:
@rainers - Yes, I did indeed have a stronger condition, I'll give it a test with `mangled == NULL || !dlang_symbol_name_p (mangled, info)` instead.

Reply:
In answer, passing tests just fail.  So how it is now is the best protection for it (dlang_parse_qualified should never parse a symbol to its end).

Reply:
If it hits a non-zero character (even an expected one), it could still be wrong and just be a coincidence from where ever the type decoding happens to end.

IIRC the "nofunction" version was modelled according to a conflict free LALR1 grammar of the mangling (https://gist.githubusercontent.com/rainers/6cdf73b48837defb9f88/raw/7ded908d44df10258e1be82d662a2ab4b82074d4/dwebsite5.bison) by using some special tokens that have a larger lookahead.

Reply:
The test that fails:
```
_D8demangle32__T2fnTS3symVS3valS1a4_6e756c6cZ3fun13__T8positionZ13__T8confusesZ8demangleFDFxaZvZv
```
Unfortunately, I don't remember how much of that symbol is real.  I do seem to recall trying various signatures to force the compiler to emit the above ambiguous symbol.

TypeNoFunction is not really implemented properly in this demangler, as the casing for type functions remains in what was `dlang_type_nofunction`.

I'll have to dump all symbols from phobos as a test and run them through this implementation to see what symbols in the wild are unparsable.

Reply:
Apart from `sym` and `val` having an additional module name, this reproduces the symbol with a compiler prior to dmd 2.077 (that added back references and removed the template length):
```D
module demangle;
struct sym {}
struct val { string s; }

template fn(T, val V)
{
   struct fun
   {
      template position()
      {
        template confuses()
        {
            static void demangle(void delegate(const char));
        }
      }
   }
}

void foo()
{
    fn!(sym, val("null")).fun.position!().confuses!().demangle((const char) {});
}
```
IIRC the ambiguity of the symbol was the 'V' after `3sym` that could be a Pascal function or a template value parameter. I suspect you could tweak the value so that the function argument parsing hits an 'X', 'Y' or 'Z'.